### ISSUE TITLE
Paraview gradient and compute_gradient return different results

### ISSUE THREAD

**Issue by jrwrigh** [pyvista/pyvista-support#204]
## Description
I took an 2D UnstructuredGrid mesh, `grid`, that contains a vector array of velocities. I computed the gradient of the velocity using `grid.compute_gradient(scalars='Velocity')` and saved it to a `.vtu` file. Then I opened the file in paraview and used the "Gradient of Unstructured DataSet" filter, which uses the same `vtkGradientFilter` object. However, the resulting gradient fields are completely different:

![image](https://user-images.githubusercontent.com/20801821/87238294-f2ebdb00-c3bd-11ea-93b2-df44ef992fd4.png)

The left is the result of `compute_gradient` while the right is the result from Paraview's Filter. The result from Paraview seems _much_ closer to a realistic/correct solution than the result from the pyvista function; high values at the wall, sharp decrease away from the wall, but very similar changes across the wall. On the other hand, the `compute_gradient` result has random pockets of very large gradient far away from the wall. Note that I've changed the colorbar limits here to give a better view (the range of the data for this particular element of the gradient is -1.239E+6 to 1.3344E+6).

Additionally, the pyvista function took about 12 minutes to run on a ~165,000 point mesh, while paraview took about 10 seconds. 

I can't figure out why it took so long to run and why the end results are so different.

Paraview Version: 5.4.1
VTK Python module version: 8.2.0 
pyvista version: 0.25.3

## Example Data
[Google Drive link (23MB)](https://drive.google.com/file/d/1hj9WgRD2ENFdozxOc6TONEN7YAoXGjpT/view?usp=sharing)

**Comment by jrwrigh** [comment]
Note that paraview's python module does have a ["Gradient of Unstructured DataSet"](https://kitware.github.io/paraview-docs/latest/python/paraview.simple.GradientOfUnstructuredDataSet.html) function documented, but I can't find any source code for it to compare with.

**Comment by banesullivan** [comment]
Are you sure you selected the right scalars in the options for the gradient filter in ParaView? I yield the same result after running the "Gradient of Unstructured DataSet" filter in ParaView with the following options (key is making sure `velocity` is also selected there)

<img width="495" alt="Screen Shot 2020-07-13 at 8 49 52 AM" src="https://user-images.githubusercontent.com/22067021/87306301-d34fd200-c4e5-11ea-8c54-6c8f593c1027.png">


Then I saved that result out to my desktop and compared to what we do in PyVista:



```py
import pyvista as pv
import numpy as np

our_grads = pv.read("testPyvista2_grad.vtu").compute_gradient(scalars='Velocity')
# Gradient of Ugrid filter with `Velocity` selected
paraview_grads = pv.read('/Users/bane/Desktop/pv_grads.vtu')

np.allclose(paraview_grads["Gradients"], our_grads["gradient"])
```
True

**Comment by jrwrigh** [comment]
I did select Velocity from the menu:
![image](https://user-images.githubusercontent.com/20801821/87317177-e2ce1b80-c4e3-11ea-9070-a4c793f7ac05.png)

I'm probably going to be pretty swamped today, but I'll try to reproduce your python test script to see if I get the same result.

I have a sneaking suspicion there might be a disconnect between VTK versions between Paraview and pyvista, but I can't figure out a way to see what VTK version Paraview is using. Though the fact that I'm using an older version of Paraview (~3 years) would suggest that old VTK is better than newer VTK for this function, which would be quite surprising.

**Comment by jrwrigh** [comment]
I got a quick opening and doing the same python script as in [your comment](https://github.com/pyvista/pyvista-support/issues/204#issuecomment-657541314) I get the result `False`. What version of Paraview are you using?

**Comment by banesullivan** [comment]
> I have a sneaking suspicion there might be a disconnect between VTK versions between Paraview and pyvista

Absolutely. That's an old version of ParaView. I have a feeling there must be a default filter option difference between the version of VTK bundled with ParaView and what you have in PyVista python env. To check,  open the Python shell in ParaView and import vtk and print `vtk.VTK_VERSION`


Either way, save that mesh out from ParaView and do a comparison of the actual values with PyVista to see exactly how different the results are. (e.g. compute the difference of the two arrays on the same mesh)

**Comment by banesullivan** [comment]
Welp. False isn't good

**Comment by jrwrigh** [comment]
```py
>>> import vtk
>>> vtk.VTK_VERSION
'8.1.0'
```

**Comment by banesullivan** [comment]
> would suggest that old VTK is better than newer VTK for this function, which would be quite surprising.

How can you tell/evaluate the quality of the results? What makes you think the results in your version of ParaView are better?

**Comment by banesullivan** [comment]
It does look like your version of ParaView is handling the edges better... I remember seeing something about the gradient filter changing at one point that may affect edges... not sure though

**Comment by jrwrigh** [comment]
> > would suggest that old VTK is better than newer VTK for this function, which would be quite surprising.

> How can you tell/evaluate the quality of the results? What makes you think the results in your version of ParaView are better?

Two primary ways. A) I'm quite familiar with the data and how the result should look. We have post-processing scripts that are primarily built for structured data. For idential cases (ie. only thing that changed was the mesh) they do not show the "splotches" that the pyvista gradient function is showing, ignoring the fact that the "splotches" are significantly higher in magnitude than the gradient at the wall (which should be the highest by far). B) Just looking at the velocity field itself by eye shows that the highest change in velocity should be at the wall rather than in small pockets in the free stream.

**Comment by jrwrigh** [comment]
> Note that paraview's python module does have a "Gradient of Unstructured DataSet" function documented, but I can't find any source code for it to compare with.

Using the `inspect` module in python inside Paraview's python shell, `print(inspect.getsource(GradientOfUnstrucutredDataSet))` just returns the definition for [`CreateObject`](https://gitlab.kitware.com/paraview/paraview/-/blob/master/Wrapping/Python/paraview/simple.py#L2437), which  is just a wrapping. So there is not Python source to compare to.

**Comment by banesullivan** [comment]
I downloaded ParaView 5.4.1 and ran the gradient filter... I get the same exact result that we compute in PyVista with a newer version of VTK (i.e. the evaluation code I shared earlier is `True`).

@jrwrigh, can you send me a paraview state file with how you are computing the gradients in paraview? This will help me debug what you're doing differently

**Comment by jrwrigh** [comment]
Here's the [link to the state file](https://drive.google.com/file/d/1LnEZ-sjZLbOqG5Q78rrJEelGy8VkFw_I/view?usp=sharing). 

And here's a [link to the vtu file with the result of Paraview ('Gradients') and Pyvista ('gradient')](https://drive.google.com/file/d/1THjRst_eQ9DPKhRZrh1dBS90WmdcDqeu/view?usp=sharing). This way you can run a check to see if the result you're getting matches up with either, or is completely different and there's some other strange issue at work.

**Comment by jrwrigh** [comment]
Update: It's working now if I load the UnstructuredGrid data from the vtu file. Previously, I had been building the UnstrucutredGrid manually from desperate sources and then running `compute_gradient` on that "built" version of the grid. Let me see if I can still reproduce the issue on the "built" version of the grid.

**Comment by banesullivan** [comment]
Ah, so you weren't giving the same mesh to PyVista and ParaView?

**Comment by jrwrigh** [comment]
> Ah, so you weren't giving the same mesh to PyVista and ParaView?

Not quite. They're the same points with the same connectivity. The "builtGrid" is built from raw data points. That was then saved to a vtu file so that it could be read into Paraview. Reading that same vtu file into PyVista creates a "readGrid". 

When using the `compute_gradient` function on the builtGrid A) takes drastically longer than for the readGrid and B) computes the wrong answer. 

Using `compute_gradient` for the readGrid (be it via Paraview or PyVista) results in the correct answer and takes significantly less time.

I'm still waiting on the builtGrid.compute_gradient to finish to verify that it is still producing the incorrect result. Once I've verified that, I can pickle the object so that you can use it.

**Comment by banesullivan** [comment]
> I'm still waiting on the builtGrid.compute_gradient to finish to verify that it is still producing the incorrect result. Once I've verified that, I can pickle the object so that you can use it.

It isn't possible to pickle PyVsita meshes without going through the file reader (ref #59). You could pickle the numpy arrays and send me the code to create the grid from those arrays though

**Comment by jrwrigh** [comment]
I've just packaged up a full example script. Didn't take too long from what I've already written. 
https://drive.google.com/drive/folders/17vDEGPOl9YTg-Ay9TlkLZexmcBXeHP_S?usp=sharing

The `.crd` file holds the actual point locations. I didn't have a connectivity file for this mesh, so I had to generate one myself using  `scipy.spatial.Delaunay`. The data comes from the `velbar.10000.1` file, which is in a binary format from Fortran. The grid is built from that data. I've commented out the `builtGridGrad = builtGrid.compute_gradient(scalars='Velocity')` since it takes awhile to run. Instead, I load a previous vtu file that contains the gradient data from a builtGrid.

I then do a `builtGrid.save` to save to a vtu file, then immediately read it back into `readGrid`, then `compute_gradient` on the `readGrid`. 

I think that explains everything. Let me know if you have any questions.

**Comment by banesullivan** [comment]
I think something went wrong in your creation of the `UnstructedGrid`. Honestly, I have no idea what, but I have a feeling the VTK writer is fixing the issue before saving the mesh to a file.

It may be worth using `PolyData` instead to keep things simple. Here is a rewritten version of your meshing function to return `PolyData` instead.

Also, not that offsets are no longer supported in VTK9 for UnstructuredGrids

```py
def formGrid(coordsPath, connectivityPath=None):
    """Create VTK PolyData

    Args:
        coordsPath (Path): Path to coordinates file.
        connectivityPath (Path): Path to connectivity file. If not given, will create
            mesh from the points given in the coordsPath file. (default: None)

    Returns:
        Pyvista PolyData object with the grid loaded.
    """
    coords = np.loadtxt(coordsPath)
    coords[:,2] = np.zeros(coords.shape[0])

    if not connectivityPath:
        # Generate mesh from points
        print('Generating Mesh...')
        mesh = Delaunay(coords[:,0:2])
        print('Finished Meshing!')

        nCells = mesh.simplices.shape[0]
            # mesh.simplices contains the connectivity array
        connectivity = mesh.simplices
        connectivity = np.hstack((np.ones((nCells,1), dtype=np.int64)*3, connectivity))

        # offsets are no longer supported in VTK9
    else:
        raise NotImplementedError("Reading in a connectivity file hasn't been coded yet! It should be pretty simple to do though.")

    return pv.PolyData(coords, connectivity)
```

with that, I believe the issue goes away.

I tried to do a point-wise comparison of the computed gradient but due to differences in the mesh, it comes back False. Instead, a visual comparison is compelling:

```py
builtGridGrad = builtGrid.compute_gradient(scalars='Velocity')
readGrid = pv.UnstructuredGrid(vtuPath.as_posix()) # the original grid you sent me
# I suspect this mesh was illformed before saving
readGridGrad = readGrid.compute_gradient(scalars='Velocity')

...

dargs = dict( scalars='gradient', log_scale=True, cmap='hot')
readGridGrad['diff'] = np.abs(builtGridGrad['gradient'] - readGridGrad['gradient'])

p = pv.Plotter(notebook=0, shape=(1,3))
p.add_mesh(builtGridGrad, **dargs)
p.subplot(0,1)
p.add_mesh(readGridGrad, **dargs)
p.subplot(0,2)
p.add_mesh(readGridGrad.copy(), scalars='diff',  cmap='magma')
p.link_views()
p.view_isometric()
p.show()
```

It appears the edges are a bit off simply due to difference in mesh quality (PloyData vs. UnstructuredGrid)

![Screen Shot 2020-07-16 at 12 10 38 AM](https://user-images.githubusercontent.com/22067021/87625735-dff84400-c6f8-11ea-86d1-2418f7b8c18e.png)

![Screen Shot 2020-07-16 at 12 10 29 AM](https://user-images.githubusercontent.com/22067021/87625741-e38bcb00-c6f8-11ea-9242-0470bf141400.png)

**Comment by jrwrigh** [comment]
> It appears the edges are a bit off simply due to difference in mesh quality (PloyData vs. UnstructuredGrid)

I'm not sure I understand. They should be the same points and connectivity, so I'm not sure where the differences in "quality" could come from. Are the connectivity arrays slightly different for some reason?

I honestly might stick with writing out a vtu file and reading it back in. It's honestly not that bad and writing out a vtu of the "raw" data anyways is pretty useful for my case.

**Comment by banesullivan** [comment]
> I'm not sure I understand. They should be the same points and connectivity, so I'm not sure where the differences in "quality" could come from. Are the connectivity arrays slightly different for some reason?

Connectivity is exactly the same. I suspect the way the ugrid is generated is wacky as the `offsets` were recently deprecated. But yeah.... I don't know why the gradient is different at the edges 🤷‍♂️ 

> I honestly might stick with writing out a vtu file and reading it back in. It's honestly not that bad and writing out a vtu of the "raw" data anyways is pretty useful for my case.

Fair enough - I like to keep things simple when possible though and `PolyData` is far simpler/lighter weight than `UnstructuredGrid`
